File: apps/cloud-extractor/app/medicine_vocabulary.py
from difflib import SequenceMatcher
import re


BRAND_ALIASES = {
    "napaone": "NapaOne",
    "napa one": "NapaOne",
    "napaoone": "NapaOne",
    "napaonee": "NapaOne",
    "napadone": "NapaOne",
    "rupa": "Rupa",
    "exium20": "Exium20",
    "exium 20": "Exium20",
    "napa extra": "Napa Extra",
    "napaextra": "Napa Extra",
    "extra": "Napa Extra",
    "trilock10": "Trilock10",
    "trilock 10": "Trilock10",
    "triloa": "Trilock10",
    "trilok10g": "Trilock10",
    "metroni": "Filmet400",
    "metrorid": "Filmet400",
    "filmet400": "Filmet400",
    "filmer400": "Filmet400",
    "norium10": "Norium 10",
    "norium1": "Norium 10",
    "norium1#": "Norium 10",
    "norium": "Norium 10",
    "oriup": "Norium 10",
    "noriup": "Norium 10",
    "coralcal-dx": "Coralcal-DX",
    "coralcal dx": "Coralcal-DX",
    "fluclox": "Fluclox",
    "cranbiotic": "Cranbiotic",
    "filfresh": "Filfresh",
    "filfresh'": "Filfresh",
}

BRAND_CANONICAL = sorted(set(BRAND_ALIASES.values()))

GENERIC_CANONICAL = {
    "paracetamol": "Paracetamol",
    "esomeprazole": "Esomeprazole",
    "rupatadine": "Rupatadine",
    "metronidazole": "Metronidazole",
    "melatonin": "Melatonin",
    "flucloxacillin": "Flucloxacillin",
    "montelukast": "Montelukast",
    "flunarizine": "Flunarizine",
    "cranberry": "Cranberry",
    "calcium": "Calcium",
    "caffeine": "Caffeine",
}


def normalize_token(value: str | None) -> str:
    return re.sub(r"[^a-z0-9]+", "", (value or "").strip().lower())


def correct_brand_name(value: str | None) -> str | None:
    if not value:
        return None
    lowered = value.strip().lower()
    compact = normalize_token(value)

    for alias, canonical in BRAND_ALIASES.items():
        if lowered == alias or compact == normalize_token(alias):
            return canonical

    best_match = None
    best_score = 0.0
    for alias, canonical in BRAND_ALIASES.items():
        score = SequenceMatcher(None, compact, normalize_token(alias)).ratio()
        if score > best_score:
            best_score = score
            best_match = canonical
    if best_match and best_score >= 0.82:
        return best_match
    return value


def correct_generic_line(value: str | None) -> str | None:
    if not value:
        return None

    corrected = re.sub(r"(\d+)\.m\b", r"\1 mg", value, flags=re.IGNORECASE)
    lowered = value.lower()
    for token, canonical in GENERIC_CANONICAL.items():
        variants = {token}
        compact = normalize_token(token)
        for word in re.findall(r"[a-z]+", lowered):
            if SequenceMatcher(None, normalize_token(word), compact).ratio() >= 0.8:
                variants.add(word)
        for variant in sorted(variants, key=len, reverse=True):
            corrected = re.sub(rf"\b{re.escape(variant)}\b", canonical, corrected, flags=re.IGNORECASE)

    corrected = re.sub(r"\s+", " ", corrected).strip()
    corrected = corrected.replace("&", "").strip()
    return corrected or value

File: apps/android/app/src/main/java/com/mediscan/offline/extraction/MedicineVocabulary.kt
package com.mediscan.offline.extraction

import kotlin.math.max

private val brandAliases = mapOf(
    "naprosyn500" to "Naprosyn 500",
    "naprosyn500mg" to "Naprosyn 500",
    "naprosyn" to "Naprosyn 500",
    "naprosy" to "Naprosyn 500",
    "naprosynson" to "Naprosyn 500",
    "napaone" to "NapaOne",
    "napaextra" to "Napa Extra",
    "trilock10" to "Trilock10",
    "norium10" to "Norium 10",
    "filmet400" to "Filmet400",
    "filfresh" to "Filfresh",
    "emistat" to "Emistat",
    "alatrol" to "Alatrol",
)

private val englishBrandVocabulary = listOf(
    "naprosyn",
    "napaone",
    "napa",
    "trilock",
    "norium",
    "filmet",
    "filfresh",
    "emistat",
    "alatrol",
)

private val genericCanonical = mapOf(
    "naproxen" to "Naproxen",
    "paracetamol" to "Paracetamol",
    "esomeprazole" to "Esomeprazole",
    "rupatadine" to "Rupatadine",
    "metronidazole" to "Metronidazole",
    "melatonin" to "Melatonin",
    "flucloxacillin" to "Flucloxacillin",
    "montelukast" to "Montelukast",
    "flunarizine" to "Flunarizine",
    "ondansetron" to "Ondansetron",
    "cetirizine" to "Cetirizine",
    "cranberry" to "Cranberry",
    "calcium" to "Calcium",
    "caffeine" to "Caffeine",
)

private val manufacturerAliases = mapOf(
    "radiantpharmaceuticalslimited" to "Radiant Pharmaceuticals Limited",
    "adiantphatmaceuticalslimited" to "Radiant Pharmaceuticals Limited",
    "adiantpharmaceuticalslimited" to "Radiant Pharmaceuticals Limited",
    "radiantphatmaceuticalslimited" to "Radiant Pharmaceuticals Limited",
    "apexpharmalimited" to "Apex Pharma Limited",
    "apexpharmaceuticalslimited" to "Apex Pharmaceuticals Limited",
    "beximcopharmaceuticalslimited" to "Beximco Pharmaceuticals Limited",
    "squarepharmaceuticalslimited" to "Square Pharmaceuticals Limited",
    "inceptapharmaceuticalslimited" to "Incepta Pharmaceuticals Limited",
)

fun normalizeToken(value: String?): String {
    return value
        .orEmpty()
        .lowercase()
        .replace(Regex("[^a-z0-9]+"), "")
}

fun correctBrandName(value: String?): String? {
    if (value.isNullOrBlank()) {
        return null
    }

    val compact = normalizeToken(value)
    brandAliases[compact]?.let { return it }

    var bestMatch: String? = null
    var bestDistance = Int.MAX_VALUE
    for ((alias, canonical) in brandAliases) {
        val distance = levenshtein(compact, alias)
        if (distance < bestDistance) {
            bestDistance = distance
            bestMatch = canonical
        }
    }

    return if (bestMatch != null && bestDistance <= 2) bestMatch else value.trim()
}

fun looksLikeKnownBrand(value: String?): Boolean {
    val compact = normalizeToken(value)
    if (compact.isBlank()) {
        return false
    }
    if (brandAliases.containsKey(compact)) {
        return true
    }
    return englishBrandVocabulary.any { token ->
        compact.contains(token) || levenshtein(compact, token) <= 2
    }
}

fun correctGenericLine(value: String?): String? {
    if (value.isNullOrBlank()) {
        return null
    }

    var corrected = value
        .replace(Regex("(\\d+)\\.m\\b", RegexOption.IGNORE_CASE), "$1 mg")
        .replace(Regex("\\s+"), " ")
        .trim()

    for ((token, canonical) in genericCanonical) {
        corrected = corrected.replace(
            Regex("\\b${Regex.escape(token)}\\b", RegexOption.IGNORE_CASE),
            canonical,
        )
    }

    for ((token, canonical) in genericCanonical) {
        val words = corrected.split(" ")
        val updatedWords = words.map { word ->
            val compact = normalizeToken(word)
            if (compact.isBlank()) {
                word
            } else {
                val distance = levenshtein(compact, token)
                if (distance <= 2 && compact.length >= 5) canonical else word
            }
        }
        corrected = updatedWords.joinToString(" ").replace(Regex("\\s+"), " ").trim()
    }

    corrected = corrected.replace(Regex("""[.,;:]+$"""), "").trim()

    return corrected.ifBlank { value.trim() }
}

fun correctManufacturerName(value: String?): String? {
    if (value.isNullOrBlank()) {
        return null
    }

    val cleaned = value
        .replace(Regex("\\bphatmaceuticals\\b", RegexOption.IGNORE_CASE), "Pharmaceuticals")
        .replace(Regex("\\bradiant\\b", RegexOption.IGNORE_CASE), "Radiant")
        .replace(Regex("\\badiant\\b", RegexOption.IGNORE_CASE), "Radiant")
        .replace(Regex("\\btong\\b", RegexOption.IGNORE_CASE), "Tongi")
        .replace(Regex("\\s+"), " ")
        .trim()

    val compact = normalizeToken(cleaned)
    manufacturerAliases[compact]?.let { canonical ->
        return if (cleaned.contains(",")) {
            val suffix = cleaned.substringAfter(",", "")
            if (suffix.isBlank()) canonical else "$canonical, ${suffix.trim()}"
        } else {
            canonical
        }
    }

    var bestMatch: String? = null
    var bestDistance = Int.MAX_VALUE
    for ((alias, canonical) in manufacturerAliases) {
        val distance = levenshtein(compact, alias)
        if (distance < bestDistance) {
            bestDistance = distance
            bestMatch = canonical
        }
    }

    return if (bestMatch != null && bestDistance <= 4) {
        if (cleaned.contains(",")) {
            val suffix = cleaned.substringAfter(",", "")
            if (suffix.isBlank()) bestMatch else "$bestMatch, ${suffix.trim()}"
        } else {
            bestMatch
        }
    } else {
        cleaned
    }
}

private fun levenshtein(left: String, right: String): Int {
    if (left == right) {
        return 0
    }
    if (left.isEmpty()) {
        return right.length
    }
    if (right.isEmpty()) {
        return left.length
    }

    val costs = IntArray(right.length + 1) { it }
    for (i in 1..left.length) {
        var previousDiagonal = costs[0]
        costs[0] = i
        for (j in 1..right.length) {
            val previousAbove = costs[j]
            val substitutionCost = if (left[i - 1] == right[j - 1]) 0 else 1
            costs[j] = minOf(
                costs[j] + 1,
                costs[j - 1] + 1,
                previousDiagonal + substitutionCost,
            )
            previousDiagonal = previousAbove
        }
    }
    return costs[right.length]
}
